base/strings/faces.jl

```diff
@@ -0,0 +1,779 @@
+    end
+end
+
+"""
+    withfaces(f, kv::Pair...)
+
+Execute `f` with [`FACES`](@ref) temporarily modified by zero or more `:name =>
+val` arguments `kv`. `withfaces` is generally used via the `withfaces(kv...) do
+... end` syntax. A value of `nothing` can be used to temporarily unset an
+environment variable (if it is set). When `withfaces` returns, the original
+[`FACES`](@ref) has been restored.
+
+    !!! warning
+    Changing faces is not thread-safe.
+"""
+function withfaces(f, keyvals::Pair{Symbol, Union{Face, Symbol, Nothing}}...)
+    old = Dict{Symbol, Union{Face, Nothing}}()
+    for (name, face) in keyvals
+        old[name] = get(FACES, face, nothing)
+        if face isa Face
+            FACES[name] = face
+        elseif face isa Symbol
+            FACES[name] = get(FACES, face, Face())
+        elseif haskey(FACES, name)
+            delete!(FACES, name)
+        end
+    end
+    try f()
+    finally
+        for (name, face) in old
+            if isnothing(face)
+                delete!(FACES, name)
+            else
+                FACES[name] = face
+            end
+        end
+    end
+end
+
+"""
+    withfaces(f, altfaces::Dict{Symbol, Face})
+
+Execute `f` with [`FACES`](@ref) temporarily swapped out with `ALTFACES`
+When `withfaces` returns, the original [`FACES`](@ref) has been restored.
+
+    !!! warning
+    Changing faces is not thread-safe.
+"""
+function withfaces(f, altfaces::Dict{Symbol, Face})
+    global FACES
+    oldfaces, FACES = FACES, altfaces
+    try f()
+    finally
+        FACES = oldfaces
+    end
+end
+
+withfaces(f) = f()
+
+## Face combination and inheritance ##
+
+"""
+    merge(a::Face, b::Face)
+Merge the properties of the text faces `a` and `b`, with those
+of `b` taking priority.
+"""
+function merge(a::Face, b::Face)
+    if isempty(a.inherit)
+        Face(ifelse(isnothing(b.font),          a.font,          b.font),
+             ifelse(isnothing(b.foreground),    a.foreground,    b.foreground),
+             ifelse(isnothing(b.background),    a.background,    b.background),
+             ifelse(isnothing(b.weight),        a.weight,        b.weight),
+             ifelse(isnothing(b.slant),         a.slant,         b.slant),
+             ifelse(isnothing(b.underline),     a.underline,     b.underline),
+             ifelse(isnothing(b.strikethrough), a.strikethrough, b.strikethrough),
+             ifelse(isnothing(b.inverse),       a.inverse,       b.inverse),
+             b.inherit)
+    else
+        merge(merge(a), b)
+    end
+end
+
+"""
+    merge(faces::Face...)
+    merge(faces::Vector{Face})
+
+Merge the properties of `faces` into a single face, with the properties
+of subsequent faces overriding the properties of prior faces.
+"""
+merge(a::Face, b::Face, others::Face...) = merge(merge(a, b), others...)
+merge(faces::Vector{Face}) = foldl(merge, faces, init=Face())
```

Comment:
`merge` semantics are well defined to take `k` arguments and merge them, not to `foldl` a vector of them. The proposed `::Vector{Face}` behavior is technically allowed but I don't think it should be added.

Reply:
Suppose Base defines `merge(::Vector...)` to behave like it does for `Dict`s, returning another `Vector`. Then you're in trouble, because `merge(::Vector{Face})` is in conflict with that.

Reply:
I'll see about finding where this method is used and replacing them with the `foldl` form.

Reply:
About to push a change that removes this.